rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
-			log.Errorf("client %s failed to connect: %v", c.addr, err)
-		} else {
-			// Signal client is ready by closing Ready channel.
-			log.Infof("client %s connected", c.addr)
-			close(c.Ready)
+		<-context.Stopper.ShouldStop()
 
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
 		c.Close()
 	})
-	return c
+
+	go func() {
+		c.runHeartbeat(retryOpts)
+		c.rpcClient.Close()
+	}()
+
+	return c, nil
 }
 
-// connect dials the connection in a backoff/retry loop.
-func (c *Client) connect(opts *retry.Options, context *Context) error {
-	// Attempt to dial connection.
-	retryOpts := clientRetryOptions
-	if opts != nil {
-		retryOpts = *opts
+// Go wraps net/rpc.Client.Go and provides additional cockroach-specific error
+// handling.
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	call := c.rpcClient.Go(serviceMethod, args, reply, nil)
+	newCall := &rpc.Call{}
+	if done == nil {
+		newCall.Done = make(chan *rpc.Call, 1)
+	} else {
+		newCall.Done = done
 	}
-	retryOpts.Stopper = context.Stopper
 
-	for r := retry.Start(retryOpts); r.Next(); {
-		tlsConfig, err := context.GetClientTLSConfig()
-		if err != nil {
-			// Problem loading the TLS config. Retrying will not help.
-			return err
-		}
+	go func() {
+		finishedCall := <-call.Done
+		newCall.ServiceMethod = finishedCall.ServiceMethod
+		newCall.Args = finishedCall.Args
+		newCall.Reply = finishedCall.Reply
+		newCall.Error = finishedCall.Error
 
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				if log.V(1) {
-					log.Warning(err)
+		if newCall.Error != nil {
+			// From net/rpc/client.go; remote hung up. Our heartbeat runloop will
+			// reconnect, so this retryable.
+			if newCall.Error == io.ErrUnexpectedEOF {
+				newCall.Error = newRPCError(newCall.Error)
+			}
+		} else {
+			// Verify response data integrity if this is a proto response.
+			if resp, respOk := reply.(proto.Response); respOk {
+				if req, reqOk := args.(proto.Request); reqOk {
+					if err := resp.Verify(req); err != nil {
+						newCall.Error = err
+					}
 				}
-				continue
 			}
-			return err
-		}
-
-		c.mu.Lock()
-		c.Client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
-		c.lAddr = conn.LocalAddr()
-		c.mu.Unlock()
-		if c.lAddr == nil {
-			return errClosed
 		}
+		newCall.Done <- newCall
+	}()
 
-		// Ensure at least one heartbeat succeeds before exiting the
-		// retry loop. If it fails, don't retry: The node is probably
-		// dead.
-		if err = c.heartbeat(); err != nil {
-			return err
-		}
+	return newCall
+}
 
-		return nil
+func (c *Client) connect() error {
+	conn, err := tlsDialHTTP(c.network, c.remoteAddress, c.tlsConfig)
+	if err != nil {
+		return err
```

Comment:
We should clear `c.rpcClient` if we fail to (re-)connect, and `Go` should be updated to handle a nil rpcClient gracefully.